### PR TITLE
[BI-1473] - Append existing germplasm with optional details

### PR DESCRIPTION
**Pull request by HMS17**
# Description
**Story:** [BI-1473 - Allow user to append existing germplasm with optional details](https://breedinginsight.atlassian.net/browse/BI-1473)

A new card was created to reduce the scope of this one: https://breedinginsight.atlassian.net/browse/BI-1735?atlOrigin=eyJpIjoiYzJmNjgzZjU1YjBlNGNlNzk4YTFiYzVjZGU2YzA1NDkiLCJwIjoiaiJ9

# Dependencies
[bi-web/BI-1473](https://github.com/Breeding-Insight/bi-web/pull/265)

# Testing
see bi-web

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have tested my code and ensured it meets the acceptance criteria of the story
- [ ] I have tested that my code works with both the brapi-java-server and BreedBase
- [ ] I have create/modified unit tests to cover this change
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to documentation
- [ ] I have run TAF: _\<please include a link to TAF run>_

### CODE REVIEW

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java [hunk 1]
```diff
@@ -242,52 +259,88 @@ public Map<String, ImportPreviewStatistics> process(List<BrAPIImport> importRows
 
             Germplasm germplasm = brapiImport.getGermplasm();
 
+            // Assign the entry number
+            if (germplasm.getEntryNo() == null) {
+                germplasm.setEntryNo(Integer.toString(i + 1));
```
Comment (davedrp) [inline src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java:264]: This _could_ cause a collisions.  A user defined Entry Number for a different row may have the value i+1

**Review by davedrp — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java (hunk 1 above)
Comment (timparsons) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java:264]: I thought there was a validation to ensure that either all rows had an entry number, or no rows had an entry number?

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java [hunk 2]
```diff
@@ -242,52 +259,88 @@ public Map<String, ImportPreviewStatistics> process(List<BrAPIImport> importRows
...
+
+                    // Has an existing pedigree so add error and skip
+                    // no pedigree on second import, don't change existing
+                    // if there is pedigree & previously not, upgrade pedigree
+                    // if previous pedigree & different pedigree, validation error
+                    if (!StringUtils.isBlank(existingGermplasm.getPedigree())) {
+                        ValidationError ve = new ValidationError("Pedigree", pedigreeAlreadyExists, HttpStatus.UNPROCESSABLE_ENTITY);
+                        validationErrors.addError(i+2, ve );  // +2 instead of +1 to account for the column header row.
+                        continue;
+                    }
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java:299]: If a germplasm record has pedigree, and I include it in a second upload to append synonyms, it looks like this check would prevent that from happening.  Is that correct?

File: src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java [hunk 3]
```diff
@@ -239,14 +240,22 @@ private Map<String,BrAPIGermplasm> processGermplasmForDisplay(List<BrAPIGermplas
...
+    public List<BrAPIGermplasm> importBrAPIGermplasm(List<BrAPIGermplasm> postBrAPIGermplasmList, List<BrAPIGermplasm> putBrAPIGermplasmList, UUID programId, ImportUpload upload) throws ApiException {
         GermplasmApi api = brAPIEndpointProvider.get(programDAO.getCoreClient(programId), GermplasmApi.class);
         var program = programDAO.fetchOneById(programId);
+        Callable<Map<String, BrAPIGermplasm>> postFunction;
         try {
-            Callable<Map<String, BrAPIGermplasm>> postFunction = () -> {
-                List<BrAPIGermplasm> postResponse = brAPIDAOUtil.post(brAPIGermplasmList, upload, api::germplasmPost, importDAO::update);
-                return processGermplasmForDisplay(postResponse, program.getKey());
-            };
+            if (!postBrAPIGermplasmList.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (timparsons) [reply src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java:248]: I agree with Dave.  Probably worth creating an method like `updateBrAPIGermplasmViaImport`

File: src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java [hunk 4]
```diff
@@ -273,6 +276,15 @@ private <T, V> List<V> getListResult(ApiResponse<Pair<Optional<T>, Optional<BrAP
                 new ArrayList<>();
     }
 
+    // TODO: write generic put code
+    public <T> List<T> put(List<T> brapiObjects,
+                           ImportUpload upload,
+                           Function<List<T>, ApiResponse> putMethod,
+                           Consumer<ImportUpload> progressUpdateMethod) throws ApiException {
+
```
Comment (timparsons) [inline src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java:284]: If this isn't implemented, you may want to have this method throw an `UnsupportedOperationException` for now

**Review by timparsons — CHANGES_REQUESTED** [review]

File: src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java [hunk 5]
```diff
@@ -239,14 +240,22 @@ private Map<String,BrAPIGermplasm> processGermplasmForDisplay(List<BrAPIGermplas
         return programGermplasmMap;
     }
 
-    public List<BrAPIGermplasm> importBrAPIGermplasm(List<BrAPIGermplasm> brAPIGermplasmList, UUID programId, ImportUpload upload) throws ApiException {
+    public List<BrAPIGermplasm> importBrAPIGermplasm(List<BrAPIGermplasm> postBrAPIGermplasmList, List<BrAPIGermplasm> putBrAPIGermplasmList, UUID programId, ImportUpload upload) throws ApiException {
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java:243]: Removed

File: src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapi/v2/dao/BrAPIGermplasmDAO.java:246]: Split this method into two

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java [hunk 6]
```diff
@@ -83,17 +93,22 @@ public class GermplasmProcessor implements Processor {
     public static String duplicateEntryNoMsg = "Entry numbers must be unique. Duplicated entry numbers found: %s";
     public static String circularDependency = "Circular dependency in the pedigree tree";
     public static String listNameAlreadyExists = "Import group name already exists";
+    public static String missingGID = "No germplasm of GID %s was found in the database";
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java:96]: There are some other pre-exisiting ones that could be as well. Would you mind if they're just all kept public?

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java [hunk 7]
```diff
@@ -227,8 +243,9 @@ public Map<String, ImportPreviewStatistics> process(List<BrAPIImport> importRows
             importList = germplasm.constructBrAPIList(program, BRAPI_REFERENCE_SOURCE);
         }
 
-        // All rows are considered new germplasm, we don't check for duplicates
+        // All rows are considered new germplasm, we don't check for duplicates //todo except now we do oops
```
[the remark this answers is not in the document]
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java]: Removed

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java (hunk 1 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java:263]: @davedrp Are you ok with the duplicate check Tim referenced?

File: src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java (hunk 4 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/utilities/BrAPIDAOUtil.java]: Added `UnsupportedOperationException`

File: src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java (hunk 2 above)
Comment (nickpalladino) [reply src/main/java/org/breedinginsight/brapps/importer/services/processors/GermplasmProcessor.java]: Ya thanks, I updated the logic. 

Error conditions:
- has existing pedigree and file pedigree is different

Valid conditions:
- no existing pedigree and file different pedigree (not blank though, will fail other validations)
- existing pedigree and file pedigree same

**Review by davedrp — APPROVED** [review]

**Review by timparsons — APPROVED** [review]

**Pull request merged**